### ISSUE TITLE
a TextMate grammar editor with preview?

### ISSUE THREAD

**Issue by mickaelistria** [eclipse/tm4e#96]
Currently, editing TextMate grammar is a bit of a nightmare ;)
I would be nice if TM4E could provide a rich TextMate grammar editor, with a preview mechanism, using a "split" editor (one side edition, other side preview) or populating the Outline view with a text area showing preview, or adding a new "Preview" view.

**Comment by mickaelistria** [comment]
cc @akurtakov

**Comment by angelozerr** [comment]
You read in my mind @mickaelistria :)

Yes I would like to provide this kind of editor like PDE editor with Overview, Rules, Source page.

My idea is to develop this editor based on FormEditor to provide nicely UI like PDE and benefit with master details feature (for rule for instance). @mickaelistria what do you think?

For the moment I'm busy with  https://github.com/eclipse/tm4e/issues/94 (I'm developping an Import wizard for that, I will ping you as soon as I have something and tell you what you will think about my idea).

**Comment by mickaelistria** [comment]
The only concern I have is that if you're opting for a tab-editor, it's really annoying to switch between edition and preview with tabs. I just think the preview should be in a separate view for best results. And at the moment, I believe the preview would be more valuable for TextMate grammar developers than the editor actually.
Both could be implemented separately. I'm going to open a bug for the preview (in outline or wherever not in the editor).

**Comment by angelozerr** [comment]
I agree with you, preview tab was not in my mind. Preview should be hosted in a View (like JavaDoc view)

My idea was to displays TextMate rules in a list and when you select an item it displays rule content on the right side (like `Extension Points` PDE tab with `master detail` feature)

I think we should have a `TextMate Perspective` which opens to provide several view like http://tmtheme-editor.herokuapp.com/#!/editor/theme/Monokai:

 * a `Gallery View`  which display list of theme to select it, update it and create it. I like the 2 tabs (Light/Dark):

![image](https://cloud.githubusercontent.com/assets/1932211/25694916/4194af82-30b1-11e7-86b0-d6a4d1b7d91c.png)

 * a `Scope view`:

![image](https://cloud.githubusercontent.com/assets/1932211/25694925/4e9b6554-30b1-11e7-9018-3e788b9064df.png)

 * a `Preview view`. I tell me if we need that? I mean we could open a file (ex : *.cs file) with GenericEditor or TMEditor and see the result like:

![image](https://cloud.githubusercontent.com/assets/1932211/25695065/e8682852-30b1-11e7-8761-50a378a4b593.png)

After when you edit theme or grammar and change it, the TMPresentationReconciler could be refresed with new theme content or new grammar syntax.

What do you think about that?

**Comment by ainslec** [comment]
http://eeyo.io/iro

**Comment by RedCMD** [comment]
here's my VSCode extension to help with authoring TextMate grammars https://marketplace.visualstudio.com/items?itemName=RedCMD.tmlanguage-syntax-highlighter
